llm/ext_server/server.cpp

```diff
@@ -1382,114 +1424,161 @@ struct llama_server_context
         }
     }
 
-    void process_single_task(task_server& task)
+    std::string common_prefix(const std::string& str1, const std::string& str2) {
+        auto mismatch_pair = std::mismatch(str1.begin(), str1.end(), str2.begin());
+        return std::string(str1.begin(), mismatch_pair.first);
+    }
+
+    // Find the slot that has the greatest common prefix
+    server_slot *prefix_slot(const json &prompt) {
+        if (!prompt.is_string()) {
+            return nullptr;
+        }
+
+        std::string prompt_str = prompt.get<std::string>();
+        server_slot *slot = nullptr;
+        size_t longest = 1;
+
+        for (server_slot &s : slots) {
+            if (s.available() && s.prompt.is_string()) {
+                std::string s_prompt = s.prompt.get<std::string>();
+                std::string prefix = common_prefix(s_prompt, prompt_str);
+
+                if (prefix.size() > longest) {
+                    LOG_INFO("Slot with common prefix found", {{
+                        "slot_id", s.id,
+                        "characters", prefix.size()
+                    }});
```

Comment:
nit - doesn't this have the chance of popping out multiple times if we find later slots with longer matches?

Maybe move this to a block around line 1457 that says:
```c++
if (longest > 0)
{
  LOG_INFO("Slot with common prefix found", {{
                        "slot_id", slot.id,
                        "characters", prefix.size()
                    }});
}
```
(the prefix string would need to be defined earlier of course)

Reply:
ah shoot, thanks